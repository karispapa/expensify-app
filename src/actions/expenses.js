import uuid from 'uuid';


//ADD_EXPENSE
// action generator for add expense
export const addExpense = ({
  description = '',
  text = '',
  amount =0,
createdAt = 0}
  ={})=> ({
    type: 'ADD_EXPENSE',
    expense: {
      id: uuid(),
      description,
      text,
      amount,
      createdAt
    }
});

//REMOVE_EXPENSE
// action generator for remove expense
export const removeExpense = ({id} = {})=> ({
  type: 'REMOVE_EXPENSE',
  id
});
//EDIT_EXPENSE
// action generator for edit expense
export const editExpense = (id, updates)=> ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});