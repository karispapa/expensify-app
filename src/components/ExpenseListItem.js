import React from 'react';
import {Link} from 'react-router-dom'
// import { connect } from 'react-redux';
// import { removeExpense } from '../actions/expenses'

const ExportListItem = ({id, description, amount, createdAt})=>(

  <div>
    <Link to={`/edit/${id}`}>
     <h3>{description} </h3>
    </Link>
    
    <p> {createdAt} ksh: {amount} </p>
  </div>
);

export default ExportListItem; 