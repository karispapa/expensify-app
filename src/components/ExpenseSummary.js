import React from 'react';
import {connect} from 'react-redux';
import calculateTotals from '../selectors/expensesTotal';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral'


export class ExpenseSummary extends React.Component{
  
  render(){
    return(
      <div>
        {this.props.expenseCount === 0 ? <h1>You have no expenses</h1>:
        
            <h1>{`Viewing ${this.props.expenseCount} 
            ${this.props.expenseCount === 1 ? 'expense': 'expenses'} 
            totaling ${this.props.expenseTotal}`}</h1>
        }
      
      </div>
    )
  }
}

const mapStateToProps = (state)=>{
  const expenses = selectExpenses(state.expenses, state.filters)
  return{
  expenseCount: expenses.length,
  expenseTotal: numeral(calculateTotals(expenses)).format('$0,0.00') 
}
};

export default connect(mapStateToProps)(ExpenseSummary)