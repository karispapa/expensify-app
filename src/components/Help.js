import React from 'react';

const Help = ()=> (
  <div>
    This is the Help Page  
  </div>
);

export default Help;
