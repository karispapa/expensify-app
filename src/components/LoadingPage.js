import React from 'react';

const LoadingPage = ()=>(
  <div className="loader">
    <img className="loader__image" src="/images/loader.gif"/>
    <span className="loader__text">Please Wait</span>
  </div>
)

export default LoadingPage