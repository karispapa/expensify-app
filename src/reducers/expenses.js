

const expensesReducerDefault = [];

// Expenses reducer

export default (state = expensesReducerDefault, action)=>{

  switch(action.type){

    case 'ADD_EXPENSE':
      return [...state, action.expense]

    case 'REMOVE_EXPENSE':
      return state.filter(({id})=> action.id !== id) // filter returns all items in the array that pass the set criteria

     case 'EDIT_EXPENSE': 
     return state.map((expense)=>{
       if(expense.id === action.id){
          return {
            ...expense,
            ...action.updates
          }
       } else{
          return expense
       }
     });
    default: 
      return state
  }
};