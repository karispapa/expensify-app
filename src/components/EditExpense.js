import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {startEditExpense, startRemoveExpense} from '../actions/expenses'

export class EditExpense extends React.Component {
  // console.log(props) 
  
  onRemove = ()=>{
    this.props.startRemoveExpense(this.props.expense.id)
      this.props.history.push('/dashboard')
  };


  onSubmit = (expense)=>{
    this.props.startEditExpense(this.props.expense.id, expense)
    this.props.history.push('/dashboard')
  };

  render (){
  return (
  <div>
   <ExpenseForm 
    expense={this.props.expense}
    onSubmit={this.onSubmit}
   /> 
   <button onClick={this.onRemove}>Remove</button>
  </div>
)}};
const mapStateToProps = (state, props)=> {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
};

const mapDispatchToProps = (dispatch, props)=>{
  return{
    startEditExpense: (id, updates)=> dispatch(startEditExpense(id, updates)),
    startRemoveExpense: (id)=> dispatch(startRemoveExpense({id}))
  }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
