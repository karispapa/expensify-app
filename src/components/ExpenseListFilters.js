import React from 'react';
import { connect } from 'react-redux';
import { addTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates'
import 'react-dates/initialize';


export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }

  onDatesChange =({startDate, endDate})=>{
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate)
  };

  onFocusChange = (calenderFocused)=>{
    this.setState(()=>({calenderFocused}))
  };

  onTextChange = (e)=>{
    this.props.addTextFilter(e.target.value)
  };

  onSortChange = (e)=>{
    if(e.target.value === "date"){
      this.props.sortByDate()
    }else if(e.target.value === "amount"){
      this.props.sortByAmount()
    }
};

  render(){
    return (
      <div>
        <input type='text' value={this.props.filters.text} 
        onChange={this.onTextChange}/>  
    
        <select 
          value={this.props.filters.sortBy} // controlled input - input controlled by js
          onChange={this.onSortChange}>
    
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          onFocusChange={this.onFocusChange}
          focusedInput={this.state.calenderFocused}
          numberOfMonths={1}
          isOutsideRange={()=> false}
          showClearDates={true}
          startDateId='id1'
          endDateId='id2'
        />
        </div>
    )
  }

}

const mapStateToProps = (state)=>({ filters: state.filters })

const mapDispatchToProps = (dispatch)=>({
  addTextFilter: (text)=> dispatch(addTextFilter(text)), 
  sortByDate: ()=> dispatch(sortByDate()), 
  sortByAmount: ()=> dispatch(sortByAmount()), 
  setStartDate: (startDate)=> dispatch(setStartDate(startDate)), 
  setEndDate: (endDate)=> dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);