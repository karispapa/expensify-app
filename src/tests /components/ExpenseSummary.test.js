import React from 'react';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import {shallow} from 'enzyme'
import expenses from '../fixtures/expenses'
// import selectExpenses from '../../selectors/expenses'
import calculateTotals from '../../selectors/expensesTotal'

let expenseCount, expenseTotal, expensesArray, wrapper

beforeEach(()=>{
  expenseCount = expenses.length
  expenseTotal = calculateTotals(expenses);  

  wrapper = shallow(<ExpenseSummary  
    expenseCount={expenseCount}
    expenseTotal={expenseTotal}
    />)
})

test('Render Expense Summary correctly', ()=>{
  expect(wrapper).toMatchSnapshot()
});

test('Render Expense Summary with one expense', ()=>{
  const newExpenses = [expenses[1]]
  wrapper.setProps({
    expenseCount: newExpenses.length,
    expenseTotal: calculateTotals(newExpenses)
  })
  expect(wrapper).toMatchSnapshot()
});

test('Render Expense Summary with no expenses', ()=>{
  const newExpenses = []
  wrapper.setProps({
    expenseCount: newExpenses.length,
    expenseTotal: calculateTotals(newExpenses)
  })
  expect(wrapper).toMatchSnapshot()
});