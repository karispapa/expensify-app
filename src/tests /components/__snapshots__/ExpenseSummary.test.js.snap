// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Expense Summary correctly 1`] = `
<div>
  <h1>
    Viewing 3 
            expenses 
            totaling 115000
  </h1>
</div>
`;

exports[`Render Expense Summary with no expenses 1`] = `
<div>
  <h1>
    You have no expenses
  </h1>
</div>
`;

exports[`Render Expense Summary with one expense 1`] = `
<div>
  <h1>
    Viewing 1 
            expense 
            totaling 100000
  </h1>
</div>
`;
